netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -268,70 +270,125 @@ public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Excep
-          ctx.pipeline().replace(this, null, new SslHandler(sslEngine, false));
-        }
-      };
-      return new BufferUntilTlsNegotiatedHandler(sslBootstrap, handler);
+  static final class ClientTlsHandler extends ChannelDuplexHandler {
+
+    private final ChannelHandler next;
+    private final SslContext sslContext;
+    private final String host;
+    private final int port;
+
+    private ProtocolNegotiationEvent pne = ProtocolNegotiationEvent.DEFAULT;
+
+    ClientTlsHandler(ChannelHandler next, SslContext sslContext, String authority) {
+      this.next = checkNotNull(next, "next");
+      this.sslContext = checkNotNull(sslContext, "sslContext");
+      HostPort hostPort = parseAuthority(authority);
+      this.host = hostPort.host;
+      this.port = hostPort.port;
     }
 
     @Override
-    public AsciiString scheme() {
-      return Utils.HTTPS;
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+      SSLEngine sslEngine = sslContext.newEngine(ctx.alloc(), host, port);
+      SSLParameters sslParams = sslEngine.getSSLParameters();
+      sslParams.setEndpointIdentificationAlgorithm("HTTPS");
+      sslEngine.setSSLParameters(sslParams);
+      ctx.pipeline().addBefore(ctx.name(), /* name= */ null, new SslHandler(sslEngine, false));
+      super.handlerAdded(ctx);
     }
 
     @Override
-    public void close() {}
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+      if (evt instanceof ProtocolNegotiationEvent) {
+        pne = (ProtocolNegotiationEvent) evt;
+      } else if (evt instanceof SslHandshakeCompletionEvent) {
+        SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
+        if (handshakeEvent.isSuccess()) {
+          SslHandler handler = ctx.pipeline().get(SslHandler.class);
+          if (NEXT_PROTOCOL_VERSIONS.contains(handler.applicationProtocol())) {
+            // Successfully negotiated the protocol.
+            logSslEngineDetails(Level.FINER, ctx, "TLS negotiation succeeded.", null);
+            ctx.pipeline().replace(ctx.name(), null, next);
+            fireProtocolNegotiationEvent(ctx, handler.engine().getSession());
+          } else {
+            Exception ex = new Exception(
+                "Failed ALPN negotiation: Unable to find compatible protocol.");
+            logSslEngineDetails(Level.FINE, ctx, "TLS negotiation failed.", ex);
+            ctx.fireExceptionCaught(ex);
+          }
+        } else {
+          ctx.fireExceptionCaught(handshakeEvent.cause());
+        }
+      } else {
+        super.userEventTriggered(ctx, evt);
+      }
+    }
+
+    private void fireProtocolNegotiationEvent(ChannelHandlerContext ctx, SSLSession session) {
+      Security security = new Security(new Tls(session));
+      Attributes attrs = pne.getAttributes().toBuilder()
+          .set(GrpcAttributes.ATTR_SECURITY_LEVEL, SecurityLevel.PRIVACY_AND_INTEGRITY)
```

Comment:
couple attrs (TRANSPORT_ATTR_REMOTE_ADDR, TRANSPORT_ATTR_LOCAL_ADDR) are missing compare to BufferUntilTlsNegotiatedHandler, doesn't look like critical ones, was this intended?

Reply:
Yes.  These come from the WUAH firing an event later.   I can add them in the test.

Reply:
oh yeah, i missed that one.